clients/src/main/java/io/strimzi/http/producer/HttpProducerClient.java

```diff
@@ -0,0 +1,134 @@
+import org.apache.logging.log4j.Logger;
+
+import java.net.http.HttpClient;
+import java.net.http.HttpResponse;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+public class HttpProducerClient implements ClientsInterface {
+    private static final Logger LOGGER = LogManager.getLogger(HttpProducerClient.class);
+    private final HttpProducerConfiguration configuration;
+    private int messageIndex;
+    private int messageSuccessfullySent;
+    private HttpClient client;
+    private TracingHandle tracingHandle;
+    private HttpHandle httpHandle;
+    private final ScheduledExecutorService scheduledExecutor;
+    private CountDownLatch latch = new CountDownLatch(1);
+
+    public HttpProducerClient() {
+        this.configuration = new HttpProducerConfiguration(System.getenv());
+        this.messageIndex = 0;
+        this.client = HttpClient.newHttpClient();
+        this.tracingHandle = TracingUtil.getTracing();
+        this.httpHandle = tracingHandle.createHttpHandle("send-messages");
+        this.scheduledExecutor = Executors.newScheduledThreadPool(1, r -> new Thread(r, "http-producer"));
+    }
+
+    @Override
+    public void run() throws Exception {
+        LOGGER.info("Starting {} with configuration: \n{}", this.getClass().getName(), configuration.toString());
+        ScheduledFuture<?> future = scheduledExecutor.scheduleAtFixedRate(this::sendMessage, 0,  configuration.getDelay(), TimeUnit.MILLISECONDS);
+
+        scheduledExecutor.submit(() -> checkForCompletion(future));
+
+        latch.await();
+        scheduledExecutor.shutdown();
+
+        if (messageIndex == configuration.getMessageCount() - 1) {
+            if (messageSuccessfullySent == configuration.getMessageCount() - 1) {
+                LOGGER.info("All messages successfully sent");
+            } else {
+                LOGGER.error("Unable to correctly send all messages");
+                throw new RuntimeException("Failed to send all messages");
+            }
+        }
+    }
+
+    @Override
+    public void checkForCompletion(ScheduledFuture<?> future) {
+        if (messageIndex == configuration.getMessageCount() - 1) {
+            future.cancel(false);
+            latch.countDown();
+        }
+    }
+
+    private ProducerRecord generateMessage(int numOfMessage) {
+        String record = "{\"records\":[{\"key\":\"key-" + numOfMessage + "\",\"value\":\"" + configuration.getMessage() + "-" + numOfMessage + "\"}]}";
+
+        HttpContext context = HttpContext.post(
+            configuration.getUri(),
+            "application/vnd.kafka.json.v2+json",
+            record);
+
+        messageIndex++;
+
+        return new ProducerRecord(record, context);
+    }
+
+    public CompletableFuture<Void> sendMessage() {
```

Comment:
I am not sure this `CompletableFuture<Void>` makes sense, I mean ...
You are submitting this method to the executor service which wants a `Runnable` or `Callable`.
Why `sendMessage` can't just return `void`? What the executor service does with the `CompletableFuture` instance?
Anyway your `sendMessage` method is totally synchronous, the future is returned only when the HTTP sends ends and you get a response, there is nothing asynchronously running which complete the future but it was already returned to the caller before starting the async operation.

Reply:
Yeah, you are right -> changed it how you mentioned.